Setup.cfg config

kain88-de opened the pull request:
This enables using a setup.cfg to change build parameters of MDAnalysis. In the default version they are commented out. This will make changing these settings easier for the average user.  This would fix #462

In addition the setup.py is cleaned up a little bit. I tried to make it a little bit easier to understand.
The dependency on the shiped ez_tools is removed. (We need that to have a Py2 and Py3 compatible package anyway).

it also set's the max line length to 80 columns for flake8.

BTW should the email address given for the MAINTAINER in `setup()` be changed?

richardjgowers commented:
I didn't realise pep8 was actually 79 chars when I said 80.  We may as well just go to 79 rather than have an exception everywhere

orbeckst commented on package/setup.py:
```diff
@@ -286,27 +314,27 @@ def detect_openmp():
+                                          # and netcdf-4
               'analysis': [
                   'matplotlib',
-                  'scipy',  # sparse contact matrix
```
`scipy` is used in `analysis/psa.py`, `analysis/distances.py`,  `hbonds/hbond_autocorrel.py`, `visualization/streamlines.py` and  `visualization/streamlines_3D.py` so it should stay.

kain88-de replied:
scipy is already mentioned in the `requires` keyword. In my current understanding of the python install tools the extra_requires doesn't actually install anything so having it once in `requires` should be enough.

orbeckst replied:
At least with ezinstall there was a way to specify extras using something like `ez_install package[analysis]` in order to trigger the extra "analysis". But no-one ever used this as far as I can tell. However, if it doesn't do anything, then why even keeping `extras_requires`?

Additionally, we should actually have `'netCDF4>=1.0'` in `requires` (at least as important as `scipy`) and drop the AMBER extra.

kain88-de replied:
I'm not sure if people realize that the `analysis` module is considered an extra and not part of the core package. Especially since the analysis package is also used a lot in the documentation and has some really useful common functionality like `alignto`.  So we could say the `analysis` module is part of the core and we should require everything listed in `extras_requires` as normal installation requirements.

orbeckst replied:
We had debated this in the past and had decided that it is more important to be able to install the core with `pip install MDAnalysis` than to be guarantee that *all* the analysis modules run. The problem with scipy and matplotlib is that they don't install nicely (at least I was never able to automatically pip-install them). The same applies to netcdf (which is needed inside the "core" for AMBER). It's a compromise; I haven't heard complaints about it yet whereas I fielded lots of questions when we still had the difficult dependencies. Not everyone has the full scientific python stack installed, especially when you're fairly new to python and just heard from a colleague that "it is so easy to just do this in MDAnalysis"...

nmichaud replied:
Have you guys also considered building an anaconda package? I've found that anaconda is the easiest way to install most non-pure python packages. Disclaimer - I'm not affiliated with Continuum at all, although I know a lot of the team. Also seems like a lot of scientific folk are moving towards anaconda as their main python distro.

kain88-de replied:
Ok I understand. Then I'll change the long description to inform people that some additional dependencies are required for the full analysis package and AMBER trajectories.

The syntax to install both additional requirements would be?

    pip install 'MDAnalysis[analysis, AMBER]'

I haven't worked a lot with these extras_requires but is that the correct syntax?

tylerjereddy replied:
@nmichaud Yes, I've previously had success building an anaconda package for MDAnalysis, but interest didn't seem that high at the time.

richardjgowers replied:
We can add an anaconda package to the release process.  I've just been following the wiki page : https://github.com/MDAnalysis/mdanalysis/wiki/PreparingReleases

orbeckst replied:
@tylerjereddy � if you have some experience with anaconda package building, could you open a separate issue for building anaconda packages as part of the process (and perhaps follow up with @hainm on how they do it)?

On 2 Oct, 2015, at 07:29, Tyler Reddy wrote:

> @nmichaud Yes, I've previously had success building an anaconda package for MDAnalysis, but interest didn't seem that high at the time.
> 
> 

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

orbeckst commented on package/setup.py:
```diff
@@ -286,27 +314,27 @@ def detect_openmp():
           # all standard requirements are available through PyPi and
           # typically can be installed without difficulties through setuptools
           install_requires=[
-              'numpy>=1.5.0',  # currently not useful because without numpy we don't get here
```
I know that it seems redundant to list `numpy` here but I think we should indicate somewhere that `numpy` is needed; is there a general `requires=[...]` keyword that we could use??

kain88-de replied:
yes there is a requires keyword and that lists numpy already.

orbeckst replied:
Ok

kain88-de replied:
I also found a setuptools field that will install numpy prior to any setup or extension compilation that we are doing. I'll have a look if I can use that to have numpy installed automatically during the installation and thereby getting rid of the error message.

orbeckst commented:
Good cleanup (and good riddance to `ezinstall`), but see comments on source.

Regarding MAINTAINER email: I am not sure what you're referring to. The author address should remain  Naveen Michaud-Agrawal (= @nmichaud ); if there's a maintainer field that we haven't filled in then put Richard J Gowers (= @richardjgowers ) there.

nmichaud commented:
@orbeckst At this point I'd be happy to transition authorship to you, and just be listed as a supplemental author.

kain88-de commented:
There exists also a [maintainer field](https://docs.python.org/2/distutils/setupscript.html#additional-meta-data). So we could leave @nmichaud  as original author and enter the MDAnalysis team as maintainers with the google groups email as a contact address. I would actually like that solution.

nmichaud commented:
Have you guys also considered building an anaconda package? I've found that anaconda is the easiest way to install most non-pure python packages. Disclaimer - I'm not affiliated with Continuum at all, although I know a lot of the team. Also seems like a lot of scientific folk are moving towards anaconda as their main python distro.

hainm commented:
you guys can build mdanalysis on travis and automatically push the package to anaconda if the tests pass. No hurt to do that even if there's no one interested in now. check pytraj or mdtraj or more info. :D

orbeckst commented:
On 2 Oct, 2015, at 06:45, kain88-de wrote:

> There exists also a maintainer field. So we could leave @nmichaud as original author and enter the MDAnalysis team as maintainers with the google groups email as a contact address. I would actually like that solution.

I also like @kain88-de 's idea (unless @nmichaud objects vigorously, in which case I would put the MDAnalysis team into the author field). PyPi should continue to list @richardjgowers as the "package index owner/maintainer".

kain88-de commented:
- config wrapper (build options can be set with Environment variables or in setup.cfg). This can be used to deactivate some stuff with the pip installation eg. `MDA_USE_OPENMP=0 pip install MDAnalysis`
- changed DEBUG_CFLAGS to MDA_DEBUG_CFLAGS
- tell about analysis and AMBER extra requires in long description
- added maintainer field listing @richardjgowers 

That should be all listed complaints.

orbeckst commented on package/SUMMARY.txt:
```diff
@@ -29,6 +29,12 @@ or installed/upgraded via PyPI (using either `easy_install`_ or `pip`_)::
   easy_install -U MDAnalysis
```
We can probably delete all references to `easy_install` when we go over this the next time round. I started cleaning it from wiki and other installation instructions.

orbeckst commented:
All looking good, I am merging it now. Thanks for the clean-up!

richardjgowers commented:
I've tried putting together a release following our recipe [here](https://github.com/MDAnalysis/mdanalysis/wiki/PreparingReleases)

I got to the line `python setup.py sdist bdist bdist_egg upload` and then I get the error below.

Is this related to the changes in setup.py?

```
etc etc etc
byte-compiling build/bdist.linux-x86_64/egg/MDAnalysis/analysis/align.py to align.pyc
byte-compiling build/bdist.linux-x86_64/egg/MDAnalysis/tests/datafiles.py to datafiles.pyc
byte-compiling build/bdist.linux-x86_64/egg/MDAnalysis/tests/__init__.py to __init__.pyc
creating stub loader for MDAnalysis/coordinates/_dcdmodule.so
creating stub loader for MDAnalysis/coordinates/dcdtimeseries.so
creating stub loader for MDAnalysis/lib/_distances.so
creating stub loader for MDAnalysis/lib/_distances_openmp.so
creating stub loader for MDAnalysis/lib/parallel/distances.so
creating stub loader for MDAnalysis/lib/qcprot.so
creating stub loader for MDAnalysis/lib/_transformations.so
creating stub loader for MDAnalysis/coordinates/xdrfile/_libxdrfile2.so
Traceback (most recent call last):
  File "setup.py", line 351, in <module>
    zip_safe=False,  # as a zipped egg the *.so files are not found (at
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/usr/lib/python2.7/dist-packages/setuptools/command/bdist_egg.py", line 202, in run
    cmd.byte_compile(to_compile)
  File "/usr/lib/python2.7/distutils/command/install_lib.py", line 138, in byte_compile
    dry_run=self.dry_run)
  File "/usr/lib/python2.7/distutils/util.py", line 454, in byte_compile
    % (file, prefix))
ValueError: invalid prefix: filename 'build/bdist.linux-x86_64/egg/MDAnalysis/coordinates/_dcd.py' doesn't start with 'build/bdist.linux-x86_64/dumb'
```

orbeckst commented:
Not sure but perhaps just leave out the egg? I'd also only add `upload` once everything is exactly as it should be.

@kain88-de wanted wheel (whl) for GridDataFormats (apparently, it's the new hotness ;-) ) and I managed to produce one by installing the appropriate package. However, whl are probably more useful if we can build them on a range of systems.

kain88-de commented:
Why does it try to find a file called `_dcd.py`? I can't find such a file in the directory.

Is there a possibility to have `dry-run` uploads to check if everything works as intended? Then you can go to the commit before the merge and checkout if this PR really broke the release feature

**wheel**
I'm not sure if we can easily build wheels currently. I find it quite hard to find information about them on the web. The thing is that wheels are binary packages and that means we have to provide builds for every supported OS. I'm not sure if we currently have the hardware to do this. I think we should get some information from numpy/scipy/etc before we are investing serious time into that.

richardjgowers commented:
So I've tried it again this morning and it's worked... which actually makes me less sure of what's happening.   I'm going to do some reading to try and understand all of this.  I think we'll end up including the openmp installer fix and call it 0.12.1 instead.
